## Using DESeq2 to obtain normalized differential expression from counts generated by ht-seq count python function

###### https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html#:~:text=While%20normalization%20is%20essential%20for,counts%20between%20or%20within%20samples.

### Deseq 2 doesn' t use normalization count, rather it uses raw counts and models the normalization inside the Generalized Linear Model (GLM). These normalized counts will be useful for downstream visualization of results, but cannot be used as input to DESeq2 or any other tools that peform differential expression analysis which use the negative binomial model.

####Standard workflows like DESeq2 or edgeR will handle normalization internally, there is not much you have to do.

library(genefilter)
library(GenomicRanges)
library("DESeq2")

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("apeglm")

library("apeglm")
library("BiocParallel")
register(MulticoreParam(4))
library(RColorBrewer)
BiocManager::install("Glimma")

library(Glimma)
library(gplots)
library(org.Hs.eg.db)
library(RColorBrewer)
library(AnnotationDbi)
library(Biobase)
BiocManager::install("goseq")
BiocManager::install("geneLenDataBase")
BiocManager::install("BiasedUrn")
library(geneLenDataBase)
library(goseq)


# set working directory
#setwd("/Users/pratimachapagain/Desktop/u251_WT_IKO")
#getwd()
setwd("/Users/pratimachapagain/Desktop/u251_WT_IKO/WT_inhibitor_Vs_control")
getwd()

# Set directory where HT-seq output files exist:
#directory = ('/Users/pratimachapagain/Desktop/Deseq_Hap')


# Make sure your HT-seq count files are in the working directory, these files are generated as part of my HT-seq script


#### http://seqanswers.com/forums/showthread.php?t=35010&fbclid=IwAR0Z3lo__D2hdA3WEJ3gVOgR1R_e52U7axYFTBzqedywcZ1I49E6V_SY31M
###########################3     Feature Counts  ##############
sampleTable<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_WT_treated_Vs_control.csv", header = T)
head(sampleTable)
samples <- data.frame(row.names=c("WT_treated1","WT_treated2","WT_treated3","WT_control1", "WT_control2", "WT_control3"), condition=as.factor(c(rep("WT_treated",3),rep("WT_control",3))))
bckCDS <- DESeqDataSetFromMatrix(countData = sampleTable[,-1], colData=samples, design=~condition)

dds <- DESeq(bckCDS)
bck_res <- results(dds)
head(bck_res)

write.csv(bck_res,file="u251_WT_IPMK_inhibitor_treated_Control.csv")


############## IKO TREATED_VS_CONTROL
sampleTable<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_IKO_treated_Vs_control.csv", header = T)
head(sampleTable)
samples <- data.frame(row.names=c("IKO_treated1","IKO_treated2","IKO_treated3","IKO_control1", "IKO_control2", "IKO_control3"), condition=as.factor(c(rep("IKO_treated",3),rep("IKO_control",3))))
bckCDS <- DESeqDataSetFromMatrix(countData = sampleTable[,-1], colData=samples, design=~condition)

dds <- DESeq(bckCDS)
bck_res <- results(dds)
head(bck_res)

write.csv(bck_res,file="u251_IKO_IPMK_inhibitor_treated_Control.csv")


###### Volcano plot WT U251_TREATED_CONTROL ############
library(ggrepel)
library(ggplot2)
library(EnhancedVolcano)


#DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/WT_DE_genes_treated_control_volcano_plot.csv", header = T)
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/WT_inhibitor_Vs_control_l2fc_1.csv", header = T)
DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                
)
#drawConnectors = FALSE)



###### DE genes volcano plot u251_IKO_treated_Vs_control


DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/IKO_Inhi_treated_Vs_DMSO_volcano_plot.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                
)
#drawConnectors = FALSE)


#########   volcano plot_WT_treated_Vs_control_l2fc>1




DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_WT_IPMK_inh_Vs_control_DE_l2fc_1.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                
)
#drawConnectors = FALSE)



###### DE genes volcano plot u251_IKO_treated_Vs_control


DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_IKO_IPMK_inh_Vs_control_DE_l2fc_1.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                #xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 1, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           #1),
                #ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -7),
                
)
#drawConnectors = FALSE)














DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/WT_all_non_SIg_sig.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########


#drawConnectors = FALSE)



######### all _sig_non_sig_IKO_treated_Vs_Control


DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_all_sig_nonsig_IKO_treated_Vs_control.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                
)


DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/IKO_Inhi_treated_Vs_DMSO_volcano_plot.csv", header = T)






##########   p value scale limit ######################

DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/WT_DE_genes_treated_control_volcano_plot.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########


 ####### WT scale bar set at p value 40 ###################
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -0),
                
)
#drawConnectors = FALSE)


############   Scale limit  Pvalue =40 IKO
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/IKO_Inhi_treated_Vs_DMSO_volcano_plot.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -8),
                
)
#drawConnectors = FALSE)


####### Scale limit p value=40, l2fc>1 WTU251 #########

DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_WT_IPMK_inh_Vs_control_DE_l2fc_1.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -0),
                
)
#drawConnectors = FALSE)


####### Scale limit p value=40, l2fc>1 IKO U251 #########
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/u251_IKO_IPMK_inh_Vs_control_DE_l2fc_1.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                #xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                         #  2),
                #ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -4),
                
)
#drawConnectors = FALSE)

########################  WT l2fc_2 ################

####### Scale limit p value=40, l2fc>2 WT U251 #########
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/l2fc_2_WT_Vs_DMSO.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -4),
                
)
#drawConnectors = FALSE)


########## IKO l2Fc>2 ##############
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/l2fc_2_IKO_Vs_DMSO.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -4),
                
)
#drawConnectors = FALSE)


########## WT l2Fc>2.5 ##############
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/l2fc_2.5_WT_Vs_DMSO.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -3),
                
)
#drawConnectors = FALSE)


########## IKO l2Fc>2.5 ##############
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/l2fc_2.5_IKO_Vs_DMSO.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -1),
                
)
#drawConnectors = FALSE)



########## WT l2Fc>3 ##############
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/l2fc_3_WT_Vs_DMSO.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -3),
                
)
#drawConnectors = FALSE)


########## IKO l2Fc>3 ##############
DE_res<-read.csv("/Users/pratimachapagain/Desktop/u251_WT_IKO/l2fc_3_IKO_Vs_DMSO.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                xlim = c(min(DE_res[["log2FoldChange"]], na.rm = TRUE) - 2, max(DE_res[["log2FoldChange"]], na.rm = TRUE) +
                           2),
                ylim = c(0, max(-log10(DE_res[["padj"]]), na.rm = TRUE) -1),
                
)
#drawConnectors = FALSE)


######## GSEA 

###### https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html


#####   https://stephenturner.github.io/deseq-to-fgsea/#using_the_fgsea_package
library(tidyverse)

###  res <- read_csv("/Users/pratimachapagain/Desktop/IPMK/U251_WT_Vs_control_all_DE_genes_48hrs.csv")
res <- read_csv("/Users/pratimachapagain/Desktop/IPMK/U251_IKO_treated_Vs_control_all_DE_genes.csv")

res

res2 <- res %>% 
  dplyr::select(SYMBOL, stat) %>% 
  na.omit() %>% 
  distinct() %>% 
  group_by(SYMBOL) %>% 
  summarize(stat=mean(stat))
res2

######  Using the fgsea package
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("fgsea")
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

install.packages("ggrepel")
install.packages("RSQLite")

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("AnnotationDbi")


library(ggrepel)
library(DESeq2)
library(RSQLite)
install.packages("readxl")
install.packages("Rcpp")
library(Rcpp)
library(readxl)




library(fgsea)
ranks <- deframe(res2)
head(ranks, 20)

###### https://www.cell.com/fulltext/S2405-4712(15)00218-5 : ##########   information about hallmark pathways ############ there are overall 50  Hallmark pathways in molecular Database ###### 
pathways.hallmark <- gmtPathways("/Users/pratimachapagain/Desktop/IPMK/msigdb/h.all.v7.4.symbols.gmt")
pathways.hallmark %>% 
  head() %>% 
  lapply(head)


fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks, nperm=1000)
fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("DT")

library(DT)

fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
  arrange(padj) %>% 
  DT::datatable()


ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()


pathways.hallmark %>% 
  enframe("pathway", "SYMBOL") %>% 
  unnest() %>% 
  inner_join(res, by="SYMBOL")


gsea_output <-fgsea(pathways=gmtPathways("/Users/pratimachapagain/Desktop/IPMK/msigdb/msigdb.v7.4.symbols.gmt"), ranks, nperm=1000) %>% 
  as_tibble() %>% 
  arrange(padj)
head(gsea_output) 
#gsea_output <- apply(gsea_output,1,as.character)
typeof(gsea_output$pathway)

typeof(gsea_output$leadingEdge)
typeof(unlist(gsea_output$leadingEdge))
Leding_unl<-unlist(gsea_output$leadingEdge)
lastCol<-gsea_output$leadingEdge
#https://stackoverflow.com/questions/26564267/unlist-and-combine-elements-into-data-frame-r
lastCol1<-data.frame(colname = vapply(lastCol, function(y) {
  paste(sub("/.*", ",", y), collapse = " ")
}, character(1L)), stringsAsFactors = FALSE)


# merge data first col from data and rest from normalized data
gsea_output1<- data.frame(gsea_output[,1:7], lastCol1)
names(gsea_output1)[8] <- "leadingEdge"







#######################################  Got error    ##############################################

write.csv(gsea_output1,file="time_48hrs_WT_inhibitor_control_gsea_output.csv")
write.csv(gsea_output1,file="time_48hrs_IKO_inhibitor_control_all_genes_gsea_output.csv")
write.csv(gsea_output1,file="time_48hrs_all_genes_WT_inhibitor_control_gsea_output.csv")
write.csv(gsea_output, path="/Users/pratimachapagain/Desktop/IPMK/time_24hrs_WT_inhibitor_control_gsea_output.csv")



######### https://www.biostars.org/p/390635/       #################3 

plotEnrichment(pathway = 
                 pathways.hallmark[["SENESE_HDAC3_TARGETS_UP"]], 
               gseaParam = 1, ticksSize = 0.5, stats= ranks) 
+ labs(title="SENESE_HDAC3_TARGETS") 
+ theme(plot.title = element_text(hjust = 0.5, face="bold"))


filtered_pathway <- subset(fgseaResTidy, pval < 0.05)

filt_p <- as.vector(filtered_pathway$pathway)



for (i in filt_p){
  plotEnrichment(pathway = pathways.hallmark[[i]], 
                 gseaParam = 1, ticksSize = 0.5, stats= ranks) + 
    labs(title=i) + theme(plot.title = element_text(hjust = 0.5, face="bold"))
}




for (i in filt_p){
  pdf(paste0(i,".pdf"),height=5,width=5) # change height and width parameter
  plt <- plotEnrichment(pathway = pathways.hallmark[[i]], 
                        gseaParam = 1, ticksSize = 0.5, stats= ranks) + 
    labs(title=i) + theme(plot.title = element_text(hjust = 0.5, face="bold"))
  print(plt)
  dev.off()
}



write.csv(bck_res,file="time1_result.csv")



### In case you want to normalize your data
setwd("/Users/pratimachapagain/Desktop/IPMK")
getwd()
data<-read.csv("/Users/pratimachapagain/Desktop/IPMK/time6_2.csv", header = T)
#data<-na.omit(data)
data1<-data[,2:5]
#  3 different way to normalize the data https://www.journaldev.com/47850/normalize-data-in-r
scale_data <- as.data.frame(scale(data1))
# merge data first col from data and rest from normalized data
df = data.frame(data[,1], scale_data)
















































###### Volcano plot LN18 ############
library(ggrepel)
library(ggplot2)
library(EnhancedVolcano)

DE_res<-read.csv("/Users/pratimachapagain/Desktop/LN18/LN_18_all_sig.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                
)
#drawConnectors = FALSE)

##### LN18_l2fc_1 #########
DE_res<-read.csv("/Users/pratimachapagain/Desktop/LN18/LN18_l2fc_1.csv", header = T)

DE_res1=DE_res[,1]
EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'pvalue')

##  https://www.biostars.org/p/378955/ ########

EnhancedVolcano(DE_res,
                lab = DE_res1,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c("CRNN"),
                
)
#drawConnectors = FALSE)